### DISCUSSION TITLE
How to use OSC 52 clipboard for copy only?

### DISCUSSION THREAD

**Discussion by doums** [neovim/neovim#28010] [General]
### Problem

With terminals I use (wezterm/alacritty) the new osc52 clipboard feature works as expected for copy but not for paste operation.
For instance wezterm does not seem to allow reading the clipboard wez/wezterm/issues/2050

When I try to paste eg pressing `p` nvim become unresponsive, freeze during  ~10seconds with the following message "Waiting for OSC 52 response from the terminal. Press Ctrl-C to interrupt…".
Finally, after this 10s timeouts or pressing ctrlC, the previously yanked text is pasted.

There is a way to set `g.clipboard` to enable osc52 for copy only, and keep whatever is nvim's default for paste?

I tried this but it's not working, I lose the osc52 feature
```lua
-- this is not working
vim.g.clipboard = {
  name = 'OSC 52',
  copy = {
    ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
    ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
  },
  paste = nil -- do not override it, use nvim's default
}
```

### Steps to reproduce

1. start nvim with `nvim --clean -u mini.lua`

```mini.lua
vim.o.clipboard = 'unnamedplus'

vim.g.clipboard = {
  name = 'OSC 52',
  copy = {
    ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
    ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
  },
  paste = {
    ['+'] = require('vim.ui.clipboard.osc52').paste('+'),
    ['*'] = require('vim.ui.clipboard.osc52').paste('*'),
  },
}
```

2. yank some text eg. `y` then paste  `p`

### Expected behavior

Should past the yanked text instant and do not freeze 10s with "Waiting for OSC 52 response from the terminal. Press Ctrl-C to interrupt..." message

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2682+g011585f35f RelWithDebInfo LuaJIT 2.1.1702233742

### Vim (not Nvim) behaves the same?

-

### Operating system/version

archlinux 6.8.1-arch1-1

### Terminal name/version

- wezterm 20240203-110809-5046fc22
- alacritty 0.13.1 (fe2a3c56)

### Installation

aur neovim-git

**Comment by tristone13th** [comment]
I also encountered this problem with Windows Terminal.

**Comment by tgbrooks** [comment]
The example in my recent issue might help you (by completing disabling pasting to + and *): #28611
It also resolves a related problem in Windows Terminal (and possibly others, I haven't checked) where yanking is slow when using the `folke/which-key.nvim` plugin.

**Reply by fcying** [reply]
There are still issues with using the `no_paste` function in windows terminal.
when I paste, it returns `clipboard: provider returned invalid data`
test config
```
local g, fn = vim.g, vim.fn
g.mapleader = " "
g.mapleader = " "
g.config_dir = fn.fnamemodify(fn.resolve(fn.expand("<sfile>:p")), ":h")
g.runtime_dir = g.config_dir .. "/.repro"
for _, name in ipairs({ "config", "data", "state", "cache" }) do
    vim.env[("XDG_%s_HOME"):format(name:upper())] = g.runtime_dir .. "/" .. name
end

local lazypath = g.runtime_dir .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
    vim.fn.system({ "git", "clone", "--filter=blob:none", "--single-branch",
        "https://github.com/folke/lazy.nvim.git", lazypath })
end

vim.opt.runtimepath:prepend(lazypath)
local plugins = {
    "folke/tokyonight.nvim",
}

require("lazy").setup(plugins, {
    root = g.runtime_dir .. "/plugins",
})
vim.cmd.colorscheme("tokyonight")

vim.opt.clipboard = "unnamedplus"
function no_paste(reg)
    return function(lines)
        -- Do nothing! We can't paste with OSC52
    end
end

vim.g.clipboard = {
    name = "OSC 52",
    copy = {
         ["+"] = require("vim.ui.clipboard.osc52").copy("+"),
         ["*"] = require("vim.ui.clipboard.osc52").copy("*"),
    },
    paste = {
        ["+"] = no_paste("+"), -- Pasting disabled
        ["*"] = no_paste("*"), -- Pasting disabled
    }
}

```

**Reply by rockneverdies55** [reply]
Not sure if disabling the pasting will effect my workflow or not but this seems to fix the problem. Thank you.

**Reply by tgbrooks** [reply]
@fcying Perhaps that's due to the `-- Do nothing! We can't paste with OSC52` line since that function would normally return the value to paste. I haven't tried it but you could try returning a empty string or something there instead.

@rockneverdies55 FYI you can paste from the Windows clipboard with ctrl+shift+v even after disabling paste in `vim.g.clipboard`.

**Comment by ch4xer** [comment]
try this:

```lua
vim.api.nvim_create_autocmd('TextYankPost', {
    callback = function()
        vim.highlight.on_yank()
        local copy_to_unnamedplus = require('vim.ui.clipboard.osc52').copy('+')
        copy_to_unnamedplus(vim.v.event.regcontents)
        local copy_to_unnamed = require('vim.ui.clipboard.osc52').copy('*')
        copy_to_unnamed(vim.v.event.regcontents)
    end
})
```

while keeping your `vim.o.clipboard` as same as before, so I can still paste contents from system clipboard using `p` if there is no ssh envolved.

**Reply by ddzzkk** [reply]
This works for me.

**Reply by fcying** [reply]
There is a problem with this method, for example, when using `yy` to copy a line, the `\n` at the end is missing.

**Reply by FlameTornado10** [reply]
It seems it didn't happen when I'm using

**Reply by fcying** [reply]
> It seems it didn't happen when I'm using

This only happens when using tmux; otherwise, works fine.
```
    vim.g.clipboard = {
        name = "tmux",
        copy = {
            ["+"] = { "tmux", "load-buffer", "-" },
            ["*"] = { "tmux", "load-buffer", "-" },
        },
        paste = {
            ["+"] = { "tmux", "save-buffer", "-" },
            ["*"] = { "tmux", "save-buffer", "-" },
        },
    }
    vim.api.nvim_create_autocmd("TextYankPost", {
        group = vim.api.nvim_create_augroup("osc52", { clear = true }),
        callback = function()
            -- vim.print(vim.v.event)
            if vim.v.operator == "y" then
                require("vim.ui.clipboard.osc52").copy("+")(vim.v.event.regcontents)
                -- require("osc52").copy_register("+")
            end
        end,
    })
```

--update: use getreg to process "\n" work fine.
```
        callback = function()
            if vim.v.operator == "y" then
                local text = vim.fn.getreg("+")
                local lines = vim.split(text, "\n")
                require("vim.ui.clipboard.osc52").copy("+")(lines)
            end
        end,
```

**Reply by tadghh** [reply]
works on win term, just throw in init.lua

**Comment by ericwq** [comment]
for me, I use the following to solve the problem. no more `clipboard: provider returned invalid data`

```lua
vim.o.clipboard = "unnamedplus"

local function paste()
  return {
    vim.fn.split(vim.fn.getreg(""), "\n"),
    vim.fn.getregtype(""),
  }
end

vim.g.clipboard = {
  name = "OSC 52",
  copy = {
    ["+"] = require("vim.ui.clipboard.osc52").copy("+"),
    ["*"] = require("vim.ui.clipboard.osc52").copy("*"),
  },
  paste = {
    ["+"] = paste,
    ["*"] = paste,
  },
}
```

**Reply by fcying** [reply]
This works for me.

**Reply by CX330Blake** [reply]
works for me either

**Reply by TomBayne** [reply]
This works but i can't paste with the right-click sub-menu. Is there a way to hide that option from the menu?

**Comment by cycode0527** [comment]
my solution.
Same as [ericwq](https://github.com/ericwq)'s, but mine can separate local and remote
from <https://github.com/neovim/neovim/issues/28611>
```lua
function my_paste(reg)
    return function(lines)

        --[ 返回 “” 寄存器的内容，用来作为 p 操作符的粘贴物 ]
        local content = vim.fn.getreg('"')
        return vim.split(content, '\n')
        
    end
end

if (os.getenv('SSH_TTY') == nil)
then
    --[ 当前环境为本地环境，也包括 wsl ]
    opt.clipboard:append("unnamedplus")
else
    opt.clipboard:append("unnamedplus")
    vim.g.clipboard = {
      name = 'OSC 52',
      copy = {
        ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
        ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
      },
      paste = {
        --[ 小括号里面的内容可能是毫无意义的，但是保持原样可能看起来更好一点 ]
        ["+"] = my_paste("+"),
        ["*"] = my_paste("*"),


    },
}
end
```

**Comment by doums** [comment]
I confirm, thx to @ericwq's trick, it works for me in wezterm

```lua
local M = {}

local function paste()
  return {
    vim.split(vim.fn.getreg(''), '\n'),
    vim.fn.getregtype(''),
  }
end

function M.init()
  if vim.env.SSH_TTY then
    vim.g.clipboard = {
      name = 'OSC 52',
      copy = {
        ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
        ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
      },
      paste = {
        ['+'] = paste,
        ['*'] = paste,
      },
    }
  end
end

return M
```

**Comment by MiiKiyoshi** [comment]
```vim
" clipboard off
set clipboard=

" yank to + so OSC52 can send it
augroup ClipAuto
  autocmd!
  autocmd TextYankPost * silent! call setreg('+', getreg('"'), v:event.regtype)
augroup END

" osc52 write only (paste disabled)
lua << EOF
local osc52 = require('vim.ui.clipboard.osc52')
vim.g.clipboard = {
  name = 'osc52-write',
  copy = {
    ['+'] = osc52.copy('+'),
    ['*'] = osc52.copy('*'),
  },
  paste = {
    ['+'] = function() return { {}, 'v' } end,
    ['*'] = function() return { {}, 'v' } end,
  },
}
EOF
```

tested on windows terminal. works fine. paste with ctrl+shift+v  
no freeze, no provider error, no plugins
